test/integration/podgc/podgc_test.go

```diff
@@ -36,144 +37,181 @@ import (
+				},
+				Spec: v1.NodeSpec{},
+				Status: v1.NodeStatus{
+					Conditions: []v1.NodeCondition{
+						{
+							Type:   v1.NodeReady,
+							Status: v1.ConditionTrue,
+						},
+					},
+				},
+			}
+			node, err := cs.CoreV1().Nodes().Create(testCtx.Ctx, node, metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("Failed to create node '%v', err: %v", node.Name, err)
+			}
+
+			pod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:       "testpod",
+					Namespace:  testCtx.NS.Name,
+					Finalizers: []string{"test.k8s.io/finalizer"},
+				},
+				Spec: v1.PodSpec{
+					NodeName: node.Name,
+					Containers: []v1.Container{
+						{Name: "foo", Image: "bar"},
+					},
+				},
+			}
+
+			pod, err = cs.CoreV1().Pods(testCtx.NS.Name).Create(testCtx.Ctx, pod, metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("Error %v, while creating pod: %v", err, klog.KObj(pod))
+			}
+			defer testutils.RemovePodFinalizers(testCtx.ClientSet, t, []*v1.Pod{pod})
+			pod, err = cs.CoreV1().Pods(testCtx.NS.Name).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
+			if err != nil {
+				t.Fatalf("Error: '%v' while updating pod info: '%v'", err, klog.KObj(pod))
+			}
+
+			// we delete the node to orphan the pod
+			err = cs.CoreV1().Nodes().Delete(testCtx.Ctx, pod.Spec.NodeName, metav1.DeleteOptions{})
+			if err != nil {
+				t.Fatalf("Failed to delete node: %v, err: %v", pod.Spec.NodeName, err)
+			}
+
+			err = wait.PollImmediate(time.Second, time.Second*15, func() (bool, error) {
+				var e error
+				pod, e = cs.CoreV1().Pods(testCtx.NS.Name).Get(testCtx.Ctx, pod.Name, metav1.GetOptions{})
+				if e != nil {
+					return true, e
+				}
+				return pod.ObjectMeta.DeletionTimestamp != nil, nil
+			})
+			if err != nil {
+				t.Fatalf("Error '%v' while waiting for the pod '%v' to be deleted", err, klog.KObj(pod))
+			}
+			wantConditionType := v1.PodConditionType("TargettedForDisruption")
+			_, cond := podutil.GetPodCondition(&pod.Status, wantConditionType)
+			if test.enableJobPodFailurePolicy == true && cond == nil {
+				t.Fatalf("Pod %v/%v does not have the expected condition: %v", pod.Namespace, pod.Name, wantConditionType)
+			} else if test.enableJobPodFailurePolicy == false && cond != nil {
+				t.Fatalf("Pod %v/%v does has the unexpected condition: %v", pod.Namespace, pod.Name, wantConditionType)
+			}
+		})
 	}
 }
 
 // TestTerminatingOnOutOfServiceNode tests deletion pods terminating on out-of-service nodes
 func TestTerminatingOnOutOfServiceNode(t *testing.T) {
-	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.NodeOutOfServiceVolumeDetach, true)()
-	testCtx := setup(t, "podgc-out-of-service")
-	defer testutils.CleanupTest(t, testCtx)
-	cs := testCtx.ClientSet
-
-	node := &v1.Node{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: "node",
+	tests := map[string]struct {
+		enableJobPodFailurePolicy bool
+	}{
+		"JobPodFailurePolicy enabled": {
+			enableJobPodFailurePolicy: true,
```

Comment:
Shouldn't we test that the phase is set to Failed? And then we would need a finalizer.

Reply:
The test is extended for testing if the phase is changed with the finalizer. Still, I keep the test for the scenario when there is no finalizer to test that the pod gets deleted quickly (grace period is 0). This way I don't drop this scenario which was covered before this PR.